File: client/src/lib/atcUpdates.ts
// The ATC's own trail updates, as the ATC's word rather than as OurHike's.
//
// features/ATC_TRAIL_UPDATES.md is the design and #461 is the issue. The
// short version: the ATC publishes closures, detours and hazards at
// appalachiantrail.org/trail-updates, writes their locations in NOBO miles
// from Springer - which is what `start_mile_marker`/`end_mile_marker` already
// are - and none of it used to reach a hiker using OurHike.
//
// THE GEOMETRY NEEDED NO NEW CODE, AND THE PROVENANCE NEEDED ALL OF IT.
// `trailSlice` already turns a mile range into coordinates and `closureBands`
// already turns those into a source, so this module adapts an update into the
// shared `Closure` shape and hands it to them unchanged. What it deliberately
// does NOT do is let the adaptation reach the screen: an ATC notice, a
// verified hiker report and an unverified one are three different kinds of
// claim, and rendering them identically is the failure OurHikeValues.md #4
// exists to prevent. OurHike did not verify this. The ATC published it. Those
// are different statements, and laundering the second into the first would
// also misrepresent the ATC.
//
// So the adapter below is for GEOMETRY ONLY, and `reason_type` on its result
// is never rendered - the sheet and the banner read `AtcUpdate` directly and
// say ATC's own category in ATC's own words.
//
// WHAT IS NOT HERE. There is no live tier and there never will be: a hiker
// report has a backend endpoint behind it, and an ATC notice has ATC's
// website. `lib/publishedConditions.ts` reads the one published artifact, and
// its absence is rendered as no ATC layer rather than as a clear trail.

import type { Closure, HikeDirection } from './closureBanner'
import { isBroadAdvisory } from './closureSpan'
import { formatDistance, type UnitSystem } from './units'

/** One notice, exactly as `pipeline/export_atc_updates.py` publishes it.
 *
 *  Facts and a link, and no body text - that paragraph is ATC's writing, and
 *  `pipeline/sources.json`'s `licence` field on `atc_trail_updates` records
 *  why this stops where it does. The link is therefore load-bearing rather
 *  than a nicety: it is the whole of the detail a hiker can read. */
export interface AtcUpdate {
  /** ATC's own slug. Stable, and the only id there is - these rows never
   *  reach the closures table, so nothing ever mints a UUID for one. */
  atc_id: string
  title: string
  category: AtcCategory
  states: string[]
  start_mile_marker: number
  end_mile_marker: number
  /**
   * Whether a hiker is stopped from walking through, as the reviewer read it.
   *
   * NOT derivable from `category`, which is the thing this field replaced,
   * and ATC's live page on 2026-08-12 is why. The only notice they file as
   * `Closure` is a closed SHELTER - the trail past Limestone Spring is open -
   * while the one thing that genuinely stops a hiker, the Harpers Ferry
   * footbridge, is filed as `Detour`. A rule reading the category was wrong in
   * both directions at once: a barrier drawn across open trail for the
   * shelter, and the real obstruction included only by luck.
   *
   * So it is the reviewer's judgement, recorded per row, which is where
   * features/ATC_TRAIL_UPDATES.md puts every other judgement this data needs.
   */
  obstructs_trail: boolean
  /** ATC's `dateModified`, as an ISO string. The age a hiker cares about. */
  updated_at: string
  source_url: string
}

/** The categories ATC publishes, verbatim. Not mapped onto `ClosureReason`:
 *  that enum's labels would render an "Alert" as "Closed", which is a claim
 *  ATC did not make. */
export type AtcCategory =
  | 'Detour'
  | 'Alert'
  | 'Closure'
  | 'Parking'
  | 'Hiking Safety'
  /** Bear warnings and the like. Not among the five measured on 2026-08-09;
   *  ATC was using it for two live notices on 2026-08-12. */
  | 'Animal'

/**
 * Whether this update gets a band, rather than only a banner.
 *
 * A band's sentence is "do not walk down there, go around" - a barrier drawn
 * across the treadway. A bear warning, a closed car park and a closed shelter
 * are all real information and none of them says that, so none of them is
 * drawn as one. Getting this wrong is the same mistake `MAX_BAND_MILES` guards
 * against from the other direction: a barrier that turns out not to be a
 * barrier is what teaches a hiker that the barriers can be walked past.
 *
 * This reads the reviewer's answer rather than inferring one from ATC's
 * category - see `obstructs_trail` above for the live case that proved the
 * inference wrong.
 *
 * The undrawn ones are not dropped. They keep the banner, which needs only a
 * mile number, exactly as an over-long advisory does - so a hiker is still
 * told, in ATC's own words, that ATC has posted something about where they are
 * walking. features/ATC_TRAIL_UPDATES.md leaves "where the suppressed ones go"
 * an open question, and this is that answer for now rather than a new screen.
 */
export function obstructsTheTrail(update: AtcUpdate): boolean {
  return update.obstructs_trail
}

/** How a band's id says which update it is, and that it is ATC's.
 *
 *  Prefixed rather than bare so it can never collide with a closure's UUID in
 *  a shared id space, and so a tap that produces one of these is recognisably
 *  an ATC tap rather than an ambiguous string. */
export const ATC_BAND_ID_PREFIX = 'atc:'

export function atcBandId(update: AtcUpdate): string {
  return `${ATC_BAND_ID_PREFIX}${update.atc_id}`
}

export function atcUpdateForBandId(
  updates: readonly AtcUpdate[],
  bandId: string,
): AtcUpdate | null {
  return updates.find((update) => atcBandId(update) === bandId) ?? null
}

/**
 * An update in the shared `Closure` shape, for the geometry path only.
 *
 * `closureBands` needs an id, two mile markers and a status, and gives back
 * coordinates; reusing it means an ATC update inherits `trailSlice`'s
 * centerline placement and `isBroadAdvisory`'s length ceiling without either
 * being reimplemented, which is what #461 means by "the geometry path needs no
 * new code".
 *
 * `reason_type: 'other'` is a placeholder that must never be rendered. Its
 * label is "Closed", and applying that to an ATC Detour would put a word in
 * their mouth. Everything a hiker reads about an ATC update comes off
 * `AtcUpdate` itself - see `atcUpdateBanner` below and chrome/AtcUpdateSheet.
 *
 * `status: 'closed'` for the same narrow reason: `closureBands` drops
 * anything marked `open` because drawing a barrier across a reopened trail is
 * a false statement. A reopened ATC notice never reaches this function - the
 * reviewed file holds current notices and a person decides what "Reopened"
 * means (features/ATC_TRAIL_UPDATES.md) - so this is the honest constant
 * rather than a field with nothing behind it.
 */
export function atcUpdateAsClosure(update: AtcUpdate): Closure {
  return {
    id: atcBandId(update),
    reason_type: 'other',
    note: null,
    status: 'closed',
    start_mile_marker: update.start_mile_marker,
    end_mile_marker: update.end_mile_marker,
  }
}

/**
 * Whether ATC named a place rather than a stretch.
 *
 * Most of what they publish is a place: of the seven placeable updates live on
 * 2026-08-12, five were a single mile marker - a shelter, a footbridge, two
 * bear warnings, a flooded section. Only Hurricane Helene was a real range,
 * and it is over the band ceiling. So this is the common case, not the edge
 * one, and it is why the map draws points at all (lib/atcUpdateStyle.ts).
 */
export function isPointNotice(update: AtcUpdate): boolean {
  return update.start_mile_marker === update.end_mile_marker
}

/**
 * The updates drawn as a band along the trail, in the shared shape.
 *
 * Ranges only. A point goes to `atcPointNotices` instead, because `trailSlice`
 * would widen it to the two vertices that bracket it and draw a few dozen feet
 * of line - which is not a small band, it is an invisible one.
 */
export function atcBandCandidates(updates: readonly AtcUpdate[]): Closure[] {
  return updates
    .filter((update) => obstructsTheTrail(update) && !isPointNotice(update))
    .map(atcUpdateAsClosure)
}

/**
 * The updates drawn as a dot at a single mile.
 *
 * Unlike the bands, this is NOT limited to the ones that obstruct the trail.
 * A dot makes no claim about passability - it says "the ATC has posted
 * something here" - so a bear warning at mile 195.8 and a closed shelter at
 * mile 1,503.6 both belong on the map, and neither is the barrier a band
 * would have made them. That difference is the whole reason the two
 * geometries are decided separately rather than by one filter.
 */
export function atcPointNotices(updates: readonly AtcUpdate[]): AtcUpdate[] {
  return updates.filter(isPointNotice)
}

function mile(value: number): string {
  return value.toLocaleString('en-US', {
    minimumFractionDigits: 1,
    maximumFractionDigits: 1,
  })
}

/**
 * One line for the header, in ATC's voice.
 *
 * Deliberately not `closureBanner`'s sentence. That one opens "Trail closed",
 * which is OurHike asserting a closure; this one names the ATC first and then
 * quotes their category and their headline, so the hiker reads a claim with
 * an owner on it. The mile range is theirs too, to tenths, because that is the
 * precision they publish at.
 *
 * `null` for an update behind the hiker, and for one whose direction cannot
 * be established while they stand outside it - both rules come straight from
 * lib/closureBanner.ts and are deliberately not re-derived here: warning half
 * of all hikers about something behind them is how a warning surface teaches
 * people to ignore it.
 *
 * How far ahead it is reads in the hiker's units, like every distance in the
 * app; ATC's mile range does not, for the reason lib/units.ts gives. The two
 * sit in one sentence because they are two different facts - how far you walk
 * before you are in it, and where on the trail "it" is.
 */
export function atcUpdateBanner(
  update: AtcUpdate,
  currentMile: number,
  direction: HikeDirection | undefined,
  units: UnitSystem = 'imperial',
): string | null {
  const { start_mile_marker: start, end_mile_marker: end } = update
  const range = start === end ? `mi ${mile(start)}` : `mi ${mile(start)} – ${mile(end)}`
  const inside = currentMile >= start && currentMile <= end

  if (inside) {
    // "here" is false when "inside" spans a fifth of the trail (#485). ATC's
    // Helene advisory runs NOBO 239.4 to 637.8, so a hiker read `Alert here` for
    // 398 miles of walking - and their own text says the damage is patchy. The
    // extent replaces the position claim, in whole miles because a span this size
    // does not need tenths. ATC's category and headline are still theirs
    // verbatim; only OurHike's word for WHERE they are changes.
    const where = isBroadAtcAdvisory(update)
      ? `along ${formatDistance(Math.abs(end - start), units, 'whole')} of trail`
      : 'here'

    return `ATC · ${update.category} ${where} · ${update.title} · ${range}`
  }

  if (direction === undefined) return null

  const nearEdge = direction === 'NOBO' ? start : end
  const distanceAhead =
    direction === 'NOBO' ? nearEdge - currentMile : currentMile - nearEdge
  if (distanceAhead < 0) return null

  return `ATC · ${update.category} ${formatDistance(distanceAhead, units)} ahead · ${update.title} · ${range}`
}

/**
 * How far ahead an update is, or null if it is behind or unplaceable.
 *
 * The same rule as lib/closureBanner.ts's, and it is exported because App.tsx
 * has one banner line and two sources competing for it. Comparing distances
 * is what lets the nearer warning win, which is the rule that module already
 * applies among closures - "the closure two hundred miles north is not the one
 * that changes what they do next".
 */
export function atcUpdateDistanceAhead(
  update: AtcUpdate,
  currentMile: number,
  direction: HikeDirection | undefined,
): number | null {
  const { start_mile_marker: start, end_mile_marker: end } = update
  if (currentMile >= start && currentMile <= end) return 0
  if (direction === undefined) return null

  const nearEdge = direction === 'NOBO' ? start : end
  const ahead = direction === 'NOBO' ? nearEdge - currentMile : currentMile - nearEdge
  return ahead < 0 ? null : ahead
}

/** An update and how far ahead of the hiker it is. */
export interface RankedAtcUpdate {
  update: AtcUpdate
  distance: number
}

/** The nearest ATC update matching `wanted`, and its distance - or null. */
function nearestWhere(
  updates: readonly AtcUpdate[],
  currentMile: number,
  direction: HikeDirection | undefined,
  wanted: (update: AtcUpdate) => boolean,
): RankedAtcUpdate | null {
  let best: RankedAtcUpdate | null = null

  for (const update of updates) {
    if (!wanted(update)) continue
    const distance = atcUpdateDistanceAhead(update, currentMile, direction)
    if (distance === null) continue
    if (best === null || distance < best.distance) best = { update, distance }
  }

  return best
}

/**
 * Whether ATC named a region rather than a stretch of trail.
 *
 * Read through the shared `Closure` shape on purpose, so this is the SAME
 * ceiling `closureBands` applies rather than a second number that could drift
 * from it - the point `atcUpdateAsClosure` above is written for. Of the seven
 * placeable updates live on 2026-08-12, exactly one answered true: Helene.
 */
export function isBroadAtcAdvisory(update: AtcUpdate): boolean {
  return isBroadAdvisory(atcUpdateAsClosure(update))
}

/**
 * The two updates that matter right now, one per line of the header.
 *
 * The mirror of `closureLanes` in lib/closureBanner.ts, and it exists for the
 * same reason (#485): an update the hiker is inside scores 0 and would win the
 * urgent line outright, which is right for a footbridge and wrong for a
 * 398-mile advisory. The reasoning is written once, over there.
 *
 * Two lanes here as well rather than only on the closure side, because since
 * #461 the ATC path is the one the Helene advisory actually travels - it arrives
 * as an `AtcUpdate`, never as a `Closure`. Fixing only `closureBanner.ts` would
 * have left the exact case #485 reports still broken.
 */
export function atcUpdateLanes(
  updates: readonly AtcUpdate[],
  currentMile: number,
  direction: HikeDirection | undefined,
): { specific: RankedAtcUpdate | null; broad: RankedAtcUpdate | null } {
  return {
    specific: nearestWhere(
      updates,
      currentMile,
      direction,
      (update) => !isBroadAtcAdvisory(update),
    ),
    broad: nearestWhere(updates, currentMile, direction, isBroadAtcAdvisory),
  }
}

File: pipeline/lib/atc_updates.py
"""The shape of a reviewed ATC Trail Update, and what makes one publishable.

features/ATC_TRAIL_UPDATES.md is the design; this is the half a test can run.
It owns one question - **is this row safe to put in front of a hiker?** - and
answers it for the reviewed file (`reference/atc_updates.json`) before
`export_atc_updates.py` bakes anything.

WHY VALIDATION IS THE INTERESTING PART HERE, when the same rows in the
closures table are validated by a database. Because there is no database. ATC
updates deliberately never reach `public.closures` - `reported_by` is
`nullable=False` and an ATC notice has no reporter, and the synthetic-profile
workaround is refused by the design - so the reviewed file is edited by hand
and the only thing between a typo and a hiker is this module.

The failure that matters is a **confident wrong answer**, and it is not
hypothetical: `spike_atc_updates.py` records the near-miss where a number
pattern without the thousands separator reads `NOBO mile 1,503.6` as mile 1,
putting a Connecticut shelter in Georgia. It parses. It looks plausible. It
is 1,502 miles wrong. So the checks below are mostly about *placement being
possible at all* rather than about JSON being well-formed - a mile outside
the trail's own extent is not a location, it is a mistake with a decimal
point in it.

ONE BAD ROW FAILS THE WHOLE FILE, matching export_conditions.py's stance for
the same reason: a partial set of safety notices is worse than none, because
the gap is invisible. A hiker cannot tell a closure that was dropped from a
stretch of trail that is open.
"""

from __future__ import annotations

from datetime import date, datetime, timezone
from urllib.parse import urlparse

# The trail's own extent, from ATC's half-mile marker layer: 4,395 points,
# `Measure` running 0.5 to 2197.5. Measured against the live service on
# 2026-08-09 and repeated from spike_atc_updates.py rather than imported,
# because a spike is deliberately not a dependency of anything that ships.
#
# A parsed mile outside this is not a location. Checking it is what turns the
# thousands-separator bug from "a shelter drawn in the wrong state" into "the
# bake refused to run", and only one of those reaches a hiker.
TRAIL_MILE_MIN = 0.5
TRAIL_MILE_MAX = 2197.5

# ATC's own categories, as published on their Trail Updates page. A closed
# set: a category this does not know is a page ATC has changed the shape of,
# which is a thing to look at rather than to pass through to a hiker as an
# unrecognised word.
#
# `Animal` was not in the five measured on 2026-08-09 and was carrying two
# live bear warnings on 2026-08-12. That is this list working rather than
# failing - the reviewer met a refusal, looked, and added a word ATC had
# started using. It is the cheap version of the cost
# features/ATC_TRAIL_UPDATES.md names as "their HTML is not an API".
#
# THE SECOND ROW BELOW IS THE SAME LESSON AT SIX TIMES THE SIZE. Reviewing
# all ten pages of ATC's listing on 2026-08-24 (#945) read a category off
# every one of the 89 updates live that day, and six words appeared that the
# six above do not cover. Counts are of those 89:
#
#     Water         4    closed wells, a shelter's spigot off
#     Relocation    1    a side trail moved
#     Permits       1    the Smokies' backcountry permit system
#     Construction  1    the Bear Mtn Bridge deck replacement
#     Fire          1    Roan Mountain's burn ban, renewed to 2030
#     Conservation  1    a camping restriction in a Special Biological Area
#
# None of these is exotic and two of them - `Water` and `Fire` - sit directly
# on the paths features/../CLAUDE.md calls the ways this app can hurt
# somebody. They were invisible for as long as the review only read page one,
# which is the actual finding of #945 rather than the count of updates.
CATEGORIES = frozenset(
    {
        "Detour",
        "Alert",
        "Closure",
        "Parking",
        "Hiking Safety",
        "Animal",
        "Water",
        "Relocation",
        "Permits",
        "Construction",
        "Fire",
        "Conservation",
    }
)

# Every field a published row carries. Facts and a link - deliberately not
# ATC's body text, which is theirs (features/ATC_TRAIL_UPDATES.md, and the
# `licence` field on this source in sources.json).
REQUIRED_FIELDS = (
    "atc_id",
    "title",
    "category",
    "states",
    "start_mile_marker",
    "end_mile_marker",
    "obstructs_trail",
    "updated_at",
    "source_url",
)

#: Provenance, carried into the artifact so a display cannot outrun its
#: source (CLAUDE.md). `reviewed` means a person read ATC's page and typed the
#: row; `unreviewed` means this build parsed it off their site an hour ago and
#: nobody has checked it. Those are different claims and the app has to be
#: able to tell a hiker which one it is holding.
REVIEWED = "reviewed"
UNREVIEWED = "unreviewed"
REVIEW_STATES = frozenset({REVIEWED, UNREVIEWED})

PUBLISHED_FIELDS = (*REQUIRED_FIELDS, "review_state")


def _mile_problem(row: dict, field: str) -> str | None:
    value = row.get(field)
    if not isinstance(value, (int, float)) or isinstance(value, bool):
        return f"{field} is {value!r}, which is not a mile"
    if not TRAIL_MILE_MIN <= float(value) <= TRAIL_MILE_MAX:
        return (
            f"{field} is {value}, outside the trail's own extent "
            f"({TRAIL_MILE_MIN}-{TRAIL_MILE_MAX}). A mile off the end of the trail is a "
            "mistake with a decimal point in it, not a place - see lib/atc_updates.py."
        )
    return None


def row_problems(row: dict) -> list[str]:
    """Everything wrong with one reviewed row, in the order it was checked.

    A list rather than the first failure, so that a person fixing the file
    sees the whole of what is wrong with a row in one run instead of peeling
    it one message at a time.
    """
    problems = []

    for field in REQUIRED_FIELDS:
        if field not in row:
            problems.append(f"missing {field}")

    atc_id = row.get("atc_id")
    if "atc_id" in row and (not isinstance(atc_id, str) or not atc_id.strip()):
        problems.append("atc_id is empty")

    title = row.get("title")
    if "title" in row and (not isinstance(title, str) or not title.strip()):
        problems.append("title is empty")

    category = row.get("category")
    if "category" in row and category not in CATEGORIES:
        problems.append(f"category {category!r} is not one ATC publishes ({', '.join(sorted(CATEGORIES))})")

    states = row.get("states")
    if "states" in row and (
        not isinstance(states, list) or not states or not all(isinstance(s, str) and s.strip() for s in states)
    ):
        problems.append("states must be a non-empty list of state codes")

    for field in ("start_mile_marker", "end_mile_marker"):
        if field in row:
            problem = _mile_problem(row, field)
            if problem:
                problems.append(problem)

    start, end = row.get("start_mile_marker"), row.get("end_mile_marker")
    if isinstance(start, (int, float)) and isinstance(end, (int, float)) and not isinstance(start, bool):
        if end < start:
            # Not silently swapped. A reversed pair means the reviewer read
            # ATC's sentence backwards, and the range they meant is not
            # recoverable from the one they wrote - `closureSpanMiles` would
            # take the absolute value and draw a band that looks right while
            # having been entered wrong.
            problems.append(f"end_mile_marker {end} is before start_mile_marker {start}")

    # Required as a real boolean rather than defaulted, because the default
    # would be a guess about whether a hiker can walk through, and ATC's own
    # category cannot answer it. Measured against their live page on
    # 2026-08-12: the only notice filed as `Closure` is a closed SHELTER, with
    # the trail past it open, while the one thing that genuinely stops a hiker
    # - the Harpers Ferry footbridge - is filed as `Detour`. A rule reading
    # the category would be wrong in both directions at once.
    if "obstructs_trail" in row and not isinstance(row.get("obstructs_trail"), bool):
        problems.append(
            "obstructs_trail must be true or false - whether a hiker is stopped from walking "
            "through, which is not the same question as ATC's category (a closed shelter is a "
            "`Closure` and leaves the trail open)"
        )

    updated_at = row.get("updated_at")
    if "updated_at" in row and (not isinstance(updated_at, str) or not updated_at.strip()):
        problems.append("updated_at is empty - it is ATC's own date and the one a hiker cares about")

    source_url = row.get("source_url")
    if "source_url" in row:
        scheme = urlparse(source_url).scheme if isinstance(source_url, str) else ""
        if scheme not in ("http", "https"):
            # The same rule chrome/ClosureSheet.tsx enforces on the way out,
            # applied on the way in. A safety sheet renders this as a link a
            # hiker taps, and `javascript:` is why that validation exists;
            # refusing it here means the client's guard is a second line
            # rather than the only one.
            problems.append(f"source_url {source_url!r} is not an http(s) URL")

    return problems


def file_problems(document: dict) -> list[str]:
    """Everything wrong with the reviewed file as a whole.

    Includes the rows, and the two things only the whole file can be wrong
    about: a duplicated `atc_id`, and a review that never happened.
    """
    problems = []

    updates = document.get("updates")
    if not isinstance(updates, list):
        return ["`updates` is missing or is not a list"]

    seen = set()
    for index, row in enumerate(updates):
        if not isinstance(row, dict):
            problems.append(f"update {index} is not an object")
            continue
        label = row.get("atc_id") or f"update {index}"
        problems.extend(f"{label}: {problem}" for problem in row_problems(row))

        atc_id = row.get("atc_id")
        if isinstance(atc_id, str) and atc_id in seen:
            # The id is how the client keys a band and how a reviewer finds
            # the row again. Two rows sharing one is a copy-paste that would
            # render as one update quietly replacing another.
            problems.append(f"{atc_id}: appears more than once")
        if isinstance(atc_id, str):
            seen.add(atc_id)

    return problems


def is_reviewed(document: dict) -> bool:
    """Whether a person has actually checked this file against ATC's page.

    The file exists from the moment the feature does, with no rows in it, and
    an unreviewed empty file and a reviewed empty one are different claims:
    the first says *nobody has looked*, the second says *we looked and ATC
    has nothing placeable*. Only the second is publishable, because only the
    second is true.

    `reviewed_at` is what separates them, and it is deliberately the reviewer's
    date rather than the bake's - see export_atc_updates.py for why a daily
    bake stamping a daily `generated_at` on a three-month-old review would be
    the dishonest version of this.
    """
    reviewed_at = document.get("reviewed_at")
    return isinstance(reviewed_at, str) and bool(reviewed_at.strip())


def published_rows(document: dict) -> list[dict]:
    """The rows as the artifact carries them: the published fields, in order.

    Projected rather than passed through, so that a note a reviewer left
    themselves in the file - or ATC body text somebody pasted in while
    working - cannot reach the bucket by accident. What ships is the field
    list this module names and nothing else.
    """
    return [
        {field: (REVIEWED if field == "review_state" else row[field]) for field in PUBLISHED_FIELDS}
        for row in document["updates"]
    ]


# WHY THERE IS NO ALL-CLEAR REFUSAL HERE ANY MORE.
#
# There was one: a list of words - "reopened", "is complete", "has been
# lifted" - that refused any update whose latest edit might be announcing the
# END of the thing it warned about. #463's worry, and a fair one: an ingest
# that only ever adds notices accumulates barriers across trail people have
# been walking.
#
# It was removed on 2026-08-24 because it was wrong about live notices more
# often than it was right about dead ones. Measured against the 22 updates ATC
# had edited in the previous 90 days, it refused two, and BOTH were current:
# the Andy Layne relocation, whose "is complete" sits above a live road-walk
# on a road ATC says has "little to no shoulder", and Max Patch, whose "has
# been completed" is about restoration work inside a camping closure that runs
# to June 2029. Meanwhile the VA Creeper closure - nine miles of A.T. shut
# until March 2027 - says "will reopen", and would have been refused for it.
#
# What makes that safe to drop is that the two things #463 feared cannot
# happen here. An automatic row carries ATC's own headline verbatim, so an
# all-clear publishes AS an all-clear - "NC/TN: Iron Mtn Gap Reopened" reads
# as exactly what it is - and `auto_row` forces `obstructs_trail` false, so
# none of them can draw a barrier whatever the words say.


def _modified_after(date_modified: str, reviewed_at: str) -> bool:
    """Whether ATC edited this after the day a person last read their page."""
    try:
        edited = datetime.fromisoformat(date_modified)
    except ValueError:
        return False
    if edited.tzinfo is not None:
        edited = edited.astimezone(timezone.utc)
    try:
        looked = date.fromisoformat(reviewed_at[:10])
    except ValueError:
        return False
    return edited.date() > looked


def auto_publish_refusal(parsed, reviewed_ids: frozenset[str] | set[str], reviewed_at: str) -> str | None:
    """Why this parsed update may NOT be published without a person, or None.

    THE GATE #963 IS BUILT AROUND, and it is deliberately narrow. Everything
    it lets through has exactly one reading; everything with two readings is
    somebody's judgement and waits for one.

    THE RECENCY RULE IS THE LOAD-BEARING ONE, and it was not in #963's
    original description - it is there because the gate was run against ATC's
    real 89 updates before any of this shipped, and without it the gate
    published almost exactly the set a person had just decided to leave out.

    That is structural rather than bad luck. After a review, everything still
    unreviewed IS the reviewer's reject pile: they read all 89 and kept 35, so
    "publish what is not reviewed and parses cleanly" is a rule that publishes
    rejects. Measured 2026-08-24, without this rule the gate let through 22
    rows - twelve bear incidents from 2024 and 2025, a hunting season its own
    text ends on 2025-12-31, four vehicle break-ins, and a duplicate of a
    closure already carried.

    So an update auto-publishes only when ATC edited it AFTER the day a person
    last read their page. Then the only thing it can ever add is something
    nobody has had the chance to judge - which is the actual gap #963 exists
    to close - and it can never overturn a reviewer's decision, because every
    decision they made is about a page older than their own review date.

    Strictly after, not on-or-after: an edit landing the same day a reviewer
    worked may or may not have been seen, and the cost of waiting a day is a
    notice arriving late, against the cost of the other direction, which is
    silently republishing something a person removed on purpose.

    Returns the reason rather than a bool so the job can say what it skipped
    and why - a silent cap reads as "covered everything" when it did not.
    """
    if parsed.slug in reviewed_ids:
        # A person's row always wins. They may have given it a band, a
        # corrected mile, or decided it should not be here at all, and none of
        # those survives being overwritten by a parse.
        return "already reviewed by a person"

    if not parsed.date_modified:
        return "no dateModified on the page"
    if not _modified_after(parsed.date_modified, reviewed_at):
        return f"last edited {parsed.date_modified[:10]}, not since the review on {reviewed_at}"
    if not parsed.title or not parsed.title.strip():
        return "no title"
    if parsed.category not in CATEGORIES:
        # A word this build does not know means ATC changed the shape of their
        # page, which is a thing to look at rather than to pass through.
        return f"category {parsed.category!r} is not one this build knows"
    if not parsed.states:
        return "no states on the page"

    reference = agreed_mile(parsed.miles)
    if reference is None:
        # NOT "the first one wins". Iron Mtn Gap states five ranges
        # accumulated over months of edits and the current one is not
        # mechanically distinguishable from its own history (#463). Zero is
        # the region-wide advisory that has no place on a map at all.
        return f"{len(parsed.miles)} mile references that do not agree on one place"
    ends = [reference.start] + ([reference.end] if reference.end is not None else [])
    if any(end is None or not TRAIL_MILE_MIN <= end <= TRAIL_MILE_MAX for end in ends):
        # The thousands-separator failure, caught rather than published: a
        # mile off the end of the trail is a mistake with a decimal point in
        # it, not a place.
        return f"{reference.raw!r} is outside the trail's own extent"
    if reference.end is not None and reference.end < reference.start:
        return f"{reference.raw!r} runs backwards"

    return None


def agreed_mile(miles: list):
    """The one span every mile reference on the page agrees on, or None.

    REPEATING A MILE IS NOT AMBIGUITY, and the rule this replaced could not
    tell the difference. ATC restates a location as they edit a notice, so
    "exactly one reference" refused pages that name the same spot twice: the
    Harpers Ferry footbridge closure carries `NOBO mile 1,026.7` in both its
    2026 and 2025 sections and was held back as though the two disagreed.

    Genuine disagreement still refuses. When a notice names several DIFFERENT
    places - Iron Mtn Gap's five ranges, accumulated over months of edits -
    there is no way to tell the current one from its own history, and picking
    is a coin toss with a hiker's location.
    """
    if not miles:
        return None
    spans = {(m.start, m.end) for m in miles}
    return miles[0] if len(spans) == 1 else None


def _as_utc_stamp(iso: str) -> str:
    """ATC's `dateModified` in the `...Z` form every reviewed row already uses.

    Their JSON-LD carries a local offset (`2026-08-19T16:22:50-04:00`). A
    reviewed row carries UTC, because a person converted it by hand. Two
    spellings of the same instant in one artifact would make `updated_at`
    sortable only by accident, so the parse converts rather than passing the
    offset through.
    """
    try:
        stamped = datetime.fromisoformat(iso)
    except ValueError:
        return iso
    if stamped.tzinfo is None:
        stamped = stamped.replace(tzinfo=timezone.utc)
    return stamped.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def auto_row(parsed) -> dict:
    """One parsed update as a publishable row, with the rails bolted on.

    `obstructs_trail` IS FORCED FALSE AND IS NOT READ FROM ANYTHING. It is the
    one field that decides whether a band is drawn across the treadway, and
    `lib/atc_updates.py`'s own measurements say ATC's category cannot answer
    it - their only `Closure` on 2026-08-12 was a shelter with open trail past
    it, while the Harpers Ferry footbridge, which genuinely stops a hiker, is
    filed `Detour`. So an unreviewed row gets a dot and a banner and can never
    get a barrier. Upgrading it to one is what review is for.
    """
    reference = agreed_mile(parsed.miles)
    return {
        "atc_id": parsed.slug,
        "title": parsed.title.strip(),
        "category": parsed.category,
        "states": list(parsed.states),
        "start_mile_marker": reference.start,
        "end_mile_marker": reference.end if reference.end is not None else reference.start,
        "obstructs_trail": False,
        "updated_at": _as_utc_stamp(parsed.date_modified),
        "source_url": parsed.source_url,
        "review_state": UNREVIEWED,
    }
